src/openms/source/COMPARISON/SPECTRA/HashedSpectrum.cpp

```diff
@@ -0,0 +1,236 @@
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Lars Nilse $
+// $Authors: Lars Nilse $
+// --------------------------------------------------------------------------
+
+#include <OpenMS/KERNEL/StandardTypes.h>
+#include <OpenMS/COMPARISON/SPECTRA/HashedSpectrum.h>
+
+#include <vector>
+#include <algorithm>
+
+using namespace std;
+
+namespace OpenMS
+{
+  HashedSpectrum::HashedSpectrum(MSSpectrum<Peak1D>& raw_spectrum, double mz_bin, double mz_tolerance, bool mz_unit_ppm)
+  : mz_bin_(mz_bin), mz_tolerance_(mz_tolerance), mz_unit_ppm_(mz_unit_ppm)
+  {
+    if (raw_spectrum.size() <= 2)
+    {
+      throw Exception::InvalidSize(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, raw_spectrum.size());
+    }
+ 
+    mz_min_ = raw_spectrum.begin()->getMZ();
+      
+    std::vector<double> mz;
+    std::vector<double> intensity;
+    int last_index = 0;
+    HashedSpectrum::MzInterval bin;
+    bin.first = &(*raw_spectrum.begin());
+    bin.last = &(*raw_spectrum.begin());
+    for (MSSpectrum<Peak1D>::Iterator it = raw_spectrum.begin(); it != raw_spectrum.end(); ++it)
+    {  
+      int index = getIndex_(it->getMZ());
+  
+      if (index > last_index)
+      {
+        // close previous interval
+        bins_.push_back(bin);
+   
+        // add empty interval
+        bin.first = NULL;
+        bin.last = NULL;
+        for (int i=0; i < (index - last_index -1); ++i)
+        {
+          bins_.push_back(bin);
+        }
+    
+        // open new interval
+        bin.first = &*it;
+        bin.last = &*it;
+
+        last_index = index;
+      }
+      else
+      {
+        bin.last = &*it;
+      }
+   
+      mz.push_back(it->getMZ());
+      intensity.push_back(it->getIntensity());
+    }
+    
+    // close last interval
+    bins_.push_back(bin);              
+  }
+
+  HashedSpectrum::~HashedSpectrum()
+  {
+  }
+
+  double HashedSpectrum::getMzBin() const
+  {
+    return mz_bin_;
+  }
+  
+  double HashedSpectrum::getMzTolerance() const
+  {
+    return mz_tolerance_;
+  }
+  
+  bool HashedSpectrum::getMzUnitPpm() const
+  {
+    return mz_unit_ppm_;
+  }
+  
+  int HashedSpectrum::getIndex_(double mz)
+  {
+    if (mz_unit_ppm_)
+    {
+      // m/z intervals in ppm
+      return (int) floor(log(mz/mz_min_)/log(1.0 + mz_bin_*1.0e-6));
+    }
+    else
+    {
+      // m/z intervals in Da
+     return (int) floor((mz - mz_min_)/mz_bin_);
+    }
+  }
+  
+  double HashedSpectrum::getDistance_(double mz1, double mz2)
+  {
+    if (mz_unit_ppm_)
+    {
+      // m/z tolerance in ppm
+     return std::abs(mz1 - mz2)/mz1*1.0e+6;
```

Comment:
please check throughout the code if you can replace your calculations with functions from https://github.com/OpenMS/OpenMS/blob/develop/src/openms/include/OpenMS/MATH/MISC/MathFunctions.h